<div class="announce instapaper_body md" data-path="README.md" id="readme"><article class="markdown-body entry-content" itemprop="mainContentOfPage"><p><a href="https://travis-ci.org/bioperl/bioperl-live"><img src="https://camo.githubusercontent.com/eebe8ec1b4d6e3bfa7ac0a611fa305a1a94de6a7/68747470733a2f2f7472617669732d63692e6f72672f62696f7065726c2f62696f7065726c2d6c6976652e7376673f6272616e63683d6d6173746572" alt="Build Status" data-canonical-src="https://travis-ci.org/bioperl/bioperl-live.svg?branch=master" style="max-width:100%;"></a></p>

<h1>
<a name="user-content-getting-started" class="anchor" href="#getting-started" aria-hidden="true"><span class="octicon octicon-link"></span></a>Getting Started</h1>

<p>Please see the the <code>INSTALL</code> or <code>INSTALL.WIN</code> documents for installation
instructions.</p>

<h1>
<a name="user-content-about-bioperl" class="anchor" href="#about-bioperl" aria-hidden="true"><span class="octicon octicon-link"></span></a>About BioPerl</h1>

<p>BioPerl is a package of public domain Perl tools for computational molecular
biology.</p>

<p>Our website (<a href="http://bioperl.org/">http://bioperl.org/</a>) provides an online resource of modules,
scripts, and web links for developers of Perl-based software for life science
research.</p>

<h1>
<a name="user-content-contact-info" class="anchor" href="#contact-info" aria-hidden="true"><span class="octicon octicon-link"></span></a>Contact info</h1>

<p>BioPerl mailing list: <a href="mailto:bioperl-l@bioperl.org">bioperl-l@bioperl.org</a></p>

<p>There's quite a variety of tools available in BioPerl, and more are added all
the time. If the tool you're looking for isn't described in the documentation
please write us, it could be undocumented or in process.</p>

<ul class="task-list">
<li><p>Project website : <a href="http://bioperl.org/">http://bioperl.org/</a></p></li>
<li><p>Bug reports : <a href="https://github.com/bioperl/bioperl-live/issues">https://github.com/bioperl/bioperl-live/issues</a></p></li>
</ul><p>Please send us bugs, in particular about documentation which you think is
unclear or problems in installation. We are also very interested in functions
which don't work the way you think they do!</p>

<h1>
<a name="user-content-the-directory-structure" class="anchor" href="#the-directory-structure" aria-hidden="true"><span class="octicon octicon-link"></span></a>The directory structure</h1>

<p>The BioPerl directory structure is organized as follows:</p>

<ul class="task-list">
<li><p><strong><code>Bio/</code></strong> - BioPerl modules</p></li>
<li><p><strong><code>doc/</code></strong> - Documentation utilities</p></li>
<li><p><strong><code>examples/</code></strong> - Scripts demonstrating the many uses of BioPerl</p></li>
<li><p><strong><code>ide/</code></strong> - files for developing BioPerl using an IDE</p></li>
<li><p><strong><code>maintenance/</code></strong> - BioPerl housekeeping scripts</p></li>
<li><p><strong><code>models/</code></strong> - DIA drawing program generated OO UML for BioPerl classes
(these are quite out-of-date)</p></li>
<li><p><strong><code>scripts/</code></strong> - Useful production-quality scripts with POD documentation</p></li>
<li><p><strong><code>t/</code></strong> - Perl built-in tests, tests are divided into subdirectories
based on the specific classes being tested</p></li>
<li><p><strong><code>t/data/</code></strong> - Data files used for the tests, provides good example data</p></li>
</ul><h1>
<a name="user-content-documentation" class="anchor" href="#documentation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Documentation</h1>

<p>For documentation on BioPerl see the <strong>HOWTO</strong> documents and tutorials online at
<a href="http://bioperl.org">http://bioperl.org</a>.</p>

<p>Useful documentation in the form of example code can also be found in the
<strong><code>examples/</code></strong> and <strong><code>scripts/</code></strong> directories. The current collection includes
scripts that run BLAST, index flat files, parse PDB structure files, make
primers, retrieve ESTs based on tissue, align protein to nucleotide sequence,
run GENSCAN on multiple sequences, and much more! See <code>bioscripts.pod</code> for a
complete listing.</p>

<p>Individual <code>*.pm</code> modules have their own embedded POD documentation as well. A
complete set of hyperlinked POD, or module, documentation is available at
<a href="http://www.bioperl.org/">http://www.bioperl.org/</a>.</p>

<p>Remember that '<code>perldoc</code>' is your friend. You can use it to read any file
containing POD formatted documentation without needing any type of translator
(e.g. '<code>perldoc Bio::SeqIO</code>').</p>

<p>If you used the Build.PL installation, and depending on your platform, you may
have documentation installed as man pages, which can be accessed in the usual
way.</p>

<h1>
<a name="user-content-releases" class="anchor" href="#releases" aria-hidden="true"><span class="octicon octicon-link"></span></a>Releases</h1>

<p>BioPerl releases are always available from the website at
<a href="http://www.bioperl.org/DIST">http://www.bioperl.org/DIST</a> or in CPAN. The latest code can be found at
<a href="https://github.com/bioperl">https://github.com/bioperl</a>.</p>

<ul class="task-list">
<li>
<p>BioPerl currently uses a sematic numbering scheme to indicate stable release
series vs. development release series. A release number is a three digit
number like <code>1.2.0</code>.</p>

<ul class="task-list">
<li>The <em>first digit indicates the major release</em>, the idea being that all the
API calls in a major release are reasonably consistent.</li>
<li>The <em>second number is the release series</em>. This is probably the most
important number, and represents added functionality that is
backwards-compatible.</li>
<li>The <em>third number is the point or patch release</em> and represents mainly bug
fixes or additional code that doesn't add significant functionality to the
code base.</li>
</ul>
</li>
<li><p>From the <strong>1.0 release until the 1.6 release</strong>, even numbers (<code>1.0</code>, <code>1.2</code>, etc)
indicated stable releases. Stable releases were well tested and recommended
for most uses. Odd numbers (<code>1.1</code>, <code>1.3</code>, etc) were development releases which one
would only use if one were interested in the latest and greatest features. The
final number (e.g. <code>1.2.0</code>, <code>1.2.1</code>) is the bug fix release. The higher the number
the more bug fixes has been incorporated. In theory you can upgrade from one
bug fix release to the next with no changes to your own code (for production
cases, obviously check things out carefully before you switch over).</p></li>
<li><p>The upcoming <strong>1.7 release</strong> will be the last release series to utilize the
alternating 'stable'/'developer' convention. Starting immediately after the
final 1.6 branch, we will start splitting BioPerl into several smaller
easier-to-manage distributions. These will have independent versions, all
likely starting with v1.7.0. <strong>We do not anticipate major API changes in the
1.7.x release series*, merely that the code will be restructured in a way to
make maintenance more feasible. We anticipate retaining semantic versioning
until the **v2.x</strong> release.</p></li>
</ul><h1>
<a name="user-content-caveats-and-warnings" class="anchor" href="#caveats-and-warnings" aria-hidden="true"><span class="octicon octicon-link"></span></a>Caveats and warnings</h1>

<p>When you run the tests ("<code>./Build test</code>") some tests may issue warnings messages
or even fail. Sometimes this is because we didn't have anyone to test the test
system on the combination of your operating system, version of perl, and
associated libraries and other modules. Because BioPerl depends on several
outside libraries we may not be able to test every single combination so if
there are warnings you may find that the package is still perfectly useful.</p>

<p>If you install the bioperl-run system and run tests when you don't have the
program installed you'll get messages like '<code>program XXX not found, skipping
tests</code>'. That's okay, BioPerl is doing what it is supposed to do. If you wanted
to run the program you'd need to install it first.</p>

<p>Not all scripts in the <code>examples/</code> directory are correct and up-to-date. We need
volunteers to help maintain these so if you find they do not submit a bug report
to <a href="https://github.com/bioperl/bioperl-live/issues">https://github.com/bioperl/bioperl-live/issues</a> and consider helping out in
their maintenance.</p>

<p>If you are confused about what modules are appropriate when you try and solve a
particular issue in bioinformatics we urge you to look at HOWTO documents first.</p>

<h1>
<a name="user-content-a-simple-module-summary" class="anchor" href="#a-simple-module-summary" aria-hidden="true"><span class="octicon octicon-link"></span></a>A simple module summary</h1>

<p>Here is a quick summary of many of the useful modules and how the toolkit is
laid out:</p>

<p>All modules are in the <strong><code>Bio/</code></strong> namespace,</p>

<ul class="task-list">
<li><p><strong><code>Perl</code></strong> is for <em>new users</em>, and gives a functional interface to the main
parts of the package.</p></li>
<li>
<p><strong><code>Seq</code></strong> is for <em>Sequences</em> (protein and DNA).</p>

<ul class="task-list">
<li>
<code>Bio::PrimarySeq</code> is a plain sequence (sequence data + identifiers)</li>
<li>
<code>Bio::Seq</code> is a fancier <code>PrimarySeq</code>, in that it has annotation (via
<code>Bio::Annotation::Collection</code>) and sequence features (via <code>Bio::SeqFeatureI</code> objects, attached via
<code>Bio::FeatureHolderI</code>).</li>
<li>
<code>Bio::Seq::RichSeq</code> is all of the above, plus it has slots for extra information specific to GenBank/EMBL/SwissProt files.</li>
<li>
<code>Bio::Seq::LargeSeq</code> is for sequences which are too big for
fitting into memory.</li>
</ul>
</li>
<li><p><strong><code>SeqIO</code></strong> is for <em>reading and writing Sequences</em>. It is a front end module
for separate driver modules supporting the different sequence formats</p></li>
<li>
<p><strong><code>SeqFeature</code></strong> represent <em>start/stop/strand-based localized annotations (features) of sequences</em></p>

<ul class="task-list">
<li>
<strong><code>Bio::SeqFeature::Generic</code></strong> is basic catchall</li>
<li>
<strong><code>Bio::SeqFeature::Similarity</code></strong> a similarity sequence feature</li>
<li>
<strong><code>Bio::SeqFeature::FeaturePair</code></strong> a sequence feature which is pairwise
such as query/hit pairs</li>
</ul>
</li>
<li><p><strong><code>SearchIO</code></strong> is for <em>reading and writing pairwise alignment reports</em>, like
BLAST or FASTA</p></li>
<li>
<p><strong><code>Search</code></strong> is where the <em>alignment objects for <code>SearchIO</code> are defined</em></p>

<ul class="task-list">
<li>
<strong><code>Bio::Search::Result::GenericResult</code></strong> is the result object (a blast
query is a <code>Result</code> object)</li>
<li>
<strong><code>Bio::Search::Hit::GenericHit</code></strong> is the <code>Hit</code> object (a query will have
0 to many hits in a database)</li>
<li>
<strong><code>Bio::Search::HSP::GenericHSP</code></strong> is the High-scoring Segment Pair
object defining the alignment(s) of the query and hit.</li>
</ul>
</li>
<li><p><strong><code>SimpleAlign</code></strong> is for <em>multiple sequence alignments</em></p></li>
<li><p><strong><code>AlignIO</code></strong> is for <em>reading and writing multiple sequence alignment
formats</em></p></li>
<li><p><strong><code>Assembly</code></strong> provides the start of an <em>infrastructure for assemblies</em> and
<strong><code>Assembly::IO</code></strong> <em>IO converters</em> for them</p></li>
<li>
<p><strong><code>DB</code></strong> is the namespace for <em>all the database query classes</em></p>

<ul class="task-list">
<li>
<strong><code>Bio::DB::GenBank/GenPept</code></strong> are two modules which query NCBI entrez for
sequences</li>
<li>
<strong><code>Bio::DB::SwissProt/EMBL</code></strong> query various EMBL and SwissProt
repositories for a sequences</li>
<li>
<strong><code>Bio::DB::GFF</code></strong> is Lincoln Stein's fast, lightweight feature and
sequence database which is the backend to his GBrowse system (see
<a href="http://www.gmod.org">www.gmod.org</a>)</li>
<li>
<strong><code>Bio::DB::Flat</code></strong> is a fast implementation of the OBDA flat-file
indexing system (cross-language and cross-platform supported by O|B|F
projects see <a href="http://obda.open-bio.org">http://obda.open-bio.org</a>).</li>
<li>
<strong><code>Bio::DB::BioFetch/DBFetch</code></strong> for OBDA, Web (HTTP) access to remote
databases.</li>
<li>
<strong><code>Bio::DB::InMemoryCache/FileCache</code></strong> (fast local caching of sequences
from remote dbs to speed up your access).</li>
<li>
<strong><code>Bio::DB::Registry</code></strong> interface to the OBDA specification for remote
data sources</li>
<li>
<strong><code>Bio::DB::Biblio</code></strong> for access to remote bibliographic databases.</li>
<li>
<strong><code>Bio::DB::EUtilities</code></strong> is the initial set of modules used for generic
queried using NCBI's eUtils.</li>
</ul>
</li>
<li><p><strong><code>Annotation</code></strong> collection of <em>annotation objects</em> (comments, DBlinks,
References, and misc key/value pairs)</p></li>
<li><p><strong><code>Coordinate</code></strong> is a system for <em>mapping between different coordinate systems</em>
such as DNA to protein or between assemblies</p></li>
<li><p><strong><code>Index</code></strong> is for <em>locally indexed flatfiles</em> with BerkeleyDB</p></li>
<li>
<p><strong><code>Tools</code></strong> contains many <em>miscellaneous parsers and functions</em> for different
bioinformatics needs</p>

<ul class="task-list">
<li>Gene prediction parser (Genscan, MZEF, Grail, Genemark)</li>
<li>Annotation format (GFF)</li>
<li>Enumerate codon tables and valid sequences symbols (CodonTable,
IUPAC)</li>
<li>Phylogenetic program parsing (PAML, Molphy, Phylip)</li>
</ul>
</li>
<li><p><strong><code>Map</code></strong> represents <em>genetic and physical map representations</em></p></li>
<li><p><strong><code>Structure</code></strong> - parse and represent <em>protein structure data</em></p></li>
<li><p><strong><code>TreeIO</code></strong> is for reading and writing <em>Tree formats</em></p></li>
<li>
<p><strong><code>Tree</code></strong> is the namespace for <strong>all associated Tree classes</strong></p>

<ul class="task-list">
<li>
<strong><code>Bio::Tree::Tree</code></strong> is the basic tree object</li>
<li>
<strong><code>Bio::Tree::Node</code></strong> are the nodes which make up the tree</li>
<li>
<strong><code>Bio::Tree::Statistics</code></strong> is for computing statistics for a tree</li>
<li>
<strong><code>Bio::Tree::TreeFunctionsI</code></strong> is where specific tree functions are
implemented (like <code>is_monophyletic</code> and <code>lca</code>)</li>
</ul>
</li>
<li><p><strong><code>Bio::Biblio</code></strong> is where <em>bibliographic data and database access objects</em>
are kept</p></li>
<li><p><strong><code>Variation</code></strong> represent <em>sequences with mutations and variations</em> applied so
one can compare and represent wild-type and mutation versions of a sequence.</p></li>
<li><p><strong><code>Root</code></strong>, basic objects for the <em>internals of BioPerl</em></p></li>
</ul><h1>
<a name="user-content-upgrading-from-an-older-version" class="anchor" href="#upgrading-from-an-older-version" aria-hidden="true"><span class="octicon octicon-link"></span></a>Upgrading from an older version</h1>

<p>If you have a previously installed version of BioPerl on your system some of
these notes may help you.</p>

<ul class="task-list">
<li><p>Some modules have been removed because they have been superceded by new
development efforts. They are documented in the <strong><code>DEPRECATED</code></strong> file that is
included in the release.</p></li>
<li><p>Some methods, or the Application Programming Interface (API), have changed or
been removed. You may find that scripts which worked with BioPerl 1.4 may give
you warnings or may not work at all (although we have tried very hard to
minimize this!). Send an email to the list and we'll be happy to give you
pointers.</p></li>
</ul></article></div>